Src/FileIndexThread.pas

```diff
@@ -0,0 +1,119 @@
+unit FileIndexThread;
+
+interface
+
+uses
+  System.Classes,
+  System.SyncObjs,
+  System.SysUtils,
+  Services.IRepo;
+
+type
+  TFileIndexThread = Class(TThread)
+    strict private
+      FEvent: TEvent;
+      FRepo: IRepo;
+
+      procedure IndexFiles;
+    protected
+      procedure Execute; override;
+    public
+      constructor Create(const aRepo: IRepo); overload;
+      destructor Destroy; override;
+  End;
+
+
+implementation
+
+uses
+  ToolsAPI,
+  System.IOUtils,
+  Spring.Collections;
+
+{ TFileIndexThread }
+
+constructor TFileIndexThread.Create(const aRepo: IRepo);
+begin
+  Inherited Create(true);
+  FRepo := aRepo;
+
+  Suspended := false;
+end;
+
+destructor TFileIndexThread.Destroy;
+begin
+  FreeAndNil(FEvent);
+  inherited;
+end;
+
+procedure TFileIndexThread.Execute;
+begin
+  inherited;
+
+  FEvent := TEvent.Create(nil, True, False, '');
+  while True do begin
+
+    IndexFiles;
+
+
+    if Terminated then Break;
+    try FEvent.WaitFor(3000); except end; //Wait for 3 seconds
+    if Terminated then Break;
+  end;
+end;
```

Comment:
Unsure about the purpose of this procedure but if it needs to be there it need to be there. Seems to be a wrapper with a timeout around the IndexFiles procedure though I do find it hard to reason about what the purpose is?

Reply:
The purpose of this is rooted in the way threading in delphi works.
A thread has one functions that gets called when it starts and that is the execute function.

Here its running in a while loop to make sure it stays alive, otherwise we don't get continued updated filed.
And then the TEvent is there to make the thread wait 3 seconds after the indexing of the files, so we don't just keep doing it.